src/pegen/python_generator.py

```diff
@@ -316,9 +342,62 @@ def visit_Rhs(self, node: Rhs, is_loop: bool = False, is_gather: bool = False) -
         for alt in node.alts:
             self.visit(alt, is_loop=is_loop, is_gather=is_gather)
 
+    def print_action(
+        self,
+        action: Optional[str],
+        locations: bool,
+        unreachable: bool,
+        is_gather: bool,
+        is_loop: bool,
+        has_invalid: bool,
+    ) -> None:
+        if not action:
+            if is_gather:
+                assert len(self.local_variable_names) == 2
+                action = f"[{self.local_variable_names[0]}] + {self.local_variable_names[1]}"
+            else:
+                if has_invalid:
+                    assert unreachable
+                    assert isinstance(action, str)  # for type checker
+                elif len(self.local_variable_names) == 1:
+                    action = f"{self.local_variable_names[0]}"
+                else:
+                    action = f"[{', '.join(self.local_variable_names)}]"
+
+        if locations:
+            self.print("tok = self._tokenizer.get_last_non_whitespace_token()")
+            self.print("end_lineno, end_col_offset = tok.end")
+
+        if is_loop:
+            self.print(f"children.append({action})")
+            self.print("mark = self._mark()")
+        else:
+            self.add_return(f"{action}")
+
     def visit_Alt(self, node: Alt, is_loop: bool, is_gather: bool) -> None:
         has_cut = any(isinstance(item.item, Cut) for item in node.items)
         has_invalid = self.invalidvisitor.visit(node)
+
+        action = node.action
+        if not action and not is_gather and has_invalid:
+            action = "UNREACHABLE"
+
+        locations = False
+        unreachable = False
+        if action:
+            if "LOCATIONS" in action:
+                locations = True
+                action = action.replace("LOCATIONS", self.location_formatting)
+            if "UNREACHABLE" in action:
+                unreachable = True
+                action = action.replace("UNREACHABLE", self.unreachable_formatting)
+
+        used = None
+        if action:
+            used = self.usednamesvisitor.visit(ast.parse(action))
+            if has_cut:
+                used.add("cut")
```

Comment:
No need for the 2 ifs

```suggestion
        locations = False
        unreachable = False
        used = None
        if action:
            if "LOCATIONS" in action:
                locations = True
                action = action.replace("LOCATIONS", self.location_formatting)
            if "UNREACHABLE" in action:
                unreachable = True
                action = action.replace("UNREACHABLE", self.unreachable_formatting)

            used = self.usednamesvisitor.visit(ast.parse(action))
            if has_cut:
                used.add("cut")
```

Reply:
oh, yeah! right 😁

Reply:
Well, actually .. I think the original version is more readable. Not like the second if will lead to critical performance hit.
Here we have two if's for different logic, easy to follow and change

Reply:
Idk, I'll left it for the final review by the pablogsal

Reply:
IMO having a single if with a comment before the inner ifs and before the systematic part would be just as readable without duplicating the if test.

Reply:
If you can address this comment @0dminnimda I will merge this PR. This the last blocker on my side and I got the green light from @pablogsal

Reply:
It's not that big of a deal

Reply:
I will add some comments to the different branches and merge when I get a chance.